:lady_beetle: Handle fetching from inventory for users without a full namespaces access role

sgratch opened the pull request:
Reference: https://github.com/kubev2v/forklift-console-plugin/issues/1293

Handle providers fetching from inventory for logged users with limited namespace roles (not kubeadmin), who can only list their namespaces content and don't have get permissions to all namespaces.

Before the fix, when such a user tried to fetch providers inventory data (i.e. display the providers list), regardless to current namespace, the "`inventory server is not reachable`" error was always displayed and inventory data fetching failed.

After this fix, such a user that tries to fetch inventory data under his permitted namespace(s), will succeed to do that and no error is displayed.

## Before:
![Screenshot from 2024-08-18 14-38-01](https://github.com/user-attachments/assets/be9ed032-eb46-419d-bc0b-3b888c1ed4a4)

## After


![Screenshot from 2024-08-18 14-35-00](https://github.com/user-attachments/assets/c2bf37a1-2419-4091-9aeb-61e3b7d7a01a)

yaacov commented on packages/forklift-console-plugin/src/modules/Providers/views/list/ProvidersListPage.tsx:
```diff
@@ -179,11 +179,16 @@ const ProvidersListPage: React.FC<{
 
+  //  A list of providers belong to the active namespace
+  const currNamespaceProviders: V1beta1Provider[] = providers.filter(
+    (provider) => provider.metadata?.namespace === namespace,
```
this should not work when namespace is "all" , because each provider will have it's own namespace, for example "openshfit-mtv" and in the case of "all" we will not get this provider, but but we do want it.

so in the case we want to show only one namespace, we already have only the providers of this namespace, and we don't need this filtering.
and in the case of "all" this filter also does not add any functionality, because it will give an empty list.

I see that in `useProvidersInventoryLis` you will ignore this list in the case of "all", so we don't actually need to filter the providers in the case of "all" , and we don't need to filter them in the case of specific namespace. 

in what case do we need to filter providers by namespace ?

yaacov commented on packages/forklift-console-plugin/src/modules/Providers/hooks/useProvidersInventoryList.ts:
```diff
@@ -36,27 +38,42 @@ interface UseInventoryResult {
-          getInventoryApiUrl(`providers?detail=1`),
-        );
+        // Fetch all providers
+        if (canList) {
```
can we also pass "namespace" parameter to the `useProvidersInventoryList` method and return an error immediately in case of an "all namespace" request when the user can't list ?

yaacov commented on packages/forklift-console-plugin/src/modules/Providers/hooks/useProvidersInventoryList.ts:
```diff
@@ -36,27 +38,42 @@ interface UseInventoryResult {
+        }
+        // Fetch current namespace's providers
+        else {
+          currNamespaceProviders?.forEach(async (provider) => {
+            const newProviderInventory = await consoleFetchJSON(
+              getInventoryApiUrl(`providers/${provider.spec.type}/${provider.metadata.uid}`),
+            );
+            updateInventoryProviderIfChanged(newProviderInventory, DEFAULT_FIELDS_TO_COMPARE);
+          });
```
can you refactor this code to look similar to the "all namespace" case, e.g.

``` ts
          const newInventory: ProvidersInventoryList = await someMethodName(namespace, providers);

          updateInventoryIfChanged(newInventory, DEFAULT_FIELDS_TO_COMPARE);
```

sgratch replied:
I did use one method for both since it's more convenient and simple.

yaacov commented on packages/forklift-console-plugin/src/modules/Providers/hooks/useProvidersInventoryList.ts:
```diff
@@ -36,27 +38,42 @@ interface UseInventoryResult {
-        updateInventoryIfChanged(newInventory, DEFAULT_FIELDS_TO_COMPARE);
+          updateInventoryIfChanged(newInventory, DEFAULT_FIELDS_TO_COMPARE);
+        }
+        // Fetch current namespace's providers
```
```suggestion
        // Fetch single namespace's providers
```
you don't actually use the "current" one, you use a namespace that is passed as an argument

sgratch commented:
I addressed all of the above.
In addition, I handled the page header's error messages since it didn't work for k8s fetching errors and since we added few error types for the inventory fetching.

## Screenshots for example:
![Screenshot from 2024-08-20 17-13-43](https://github.com/user-attachments/assets/8f7cdf52-2549-4085-80bf-f64f7982db40)

![Screenshot from 2024-08-20 22-02-14](https://github.com/user-attachments/assets/78d06c7c-b8d2-41d9-84bc-c0c4ca128557)

yaacov commented on packages/forklift-console-plugin/src/modules/Providers/hooks/useProvidersInventoryList.ts:
```diff
@@ -67,7 +90,74 @@ export const useProvidersInventoryList = ({
+   * @param {string} namespace namespace to fetch providers' inventory data for.
+   * @returns {V1beta1Provider[]} list of providers' k8s data, belongs to the given namespace, or null in case of a failure.
+   */
+  function GetK8sProvidersForNamespace(namespace: string): V1beta1Provider[] {
```
it's a hook, should start with `use` and not `git`

yaacov commented on packages/forklift-console-plugin/src/modules/Providers/views/list/ProvidersListPage.tsx:
```diff
@@ -220,9 +225,9 @@ const ProvidersListPage: React.FC<{
       title={t('Providers')}
       userSettings={userSettings}
       alerts={
-        !inventoryLoading && inventoryError
```
why do we need this changes ?

sgratch replied:
Please see my following answer to next comment.

yaacov commented on packages/forklift-console-plugin/src/modules/Providers/views/list/ProvidersListPage.tsx:
```diff
@@ -239,6 +244,27 @@ const ModernizeMigration = () => (
   <img src={modernizeMigration} className="forklift-empty-state__icon" />
 );
 
+const handleK8sError = (providersLoaded: boolean, providersLoadError: Error) => {
```
why do we need this logic here ?

sgratch replied:
Before this fix, the error messages handling for displayingin page header was missing. 
There were 2 supported scenarios only:
In case of `providersLoadError`, no error msg was was  displayed in providers list page.
In case of of the `inventoryError`, only inventoryNotReachable msg was displayed.

Since this PR fetches both K8s and inventory providers resources and adds more inventory failures error types, I added this handling.
Why do you think it's  unnecessary?

yaacov commented on packages/forklift-console-plugin/src/modules/Providers/views/list/components/NamespaceNotAccessible.tsx:
why do we need this component ?

yaacov commented on packages/forklift-console-plugin/src/components/page/StandardPage.tsx:
this changes where moved to #1310

yaacov commented on packages/forklift-console-plugin/src/modules/Providers/hooks/useProvidersInventoryList.ts:
```diff
@@ -32,31 +43,55 @@ interface UseInventoryResult {
-        );
+        // Fetch all providers
+        if (canList) {
+          const newInventory: ProvidersInventoryList = await consoleFetchJSON(
```
```suggestion
          const newInventory: ProvidersInventoryList = canList
            ? await consoleFetchJSON(...
            : await getAndUpdateInventoryProviders(...
```

yaacov commented on packages/forklift-console-plugin/src/modules/Providers/hooks/useProvidersInventoryList.ts:
```diff
@@ -67,7 +102,66 @@ export const useProvidersInventoryList = ({
+      ova: [],
+    };
+
+    const allPromises = Promise.all(
+      k8Providers
+        ?.filter((provider) => provider.status.phase === 'Ready')
+        .map(async (provider) => {
+          return await consoleFetchJSON(
+            getInventoryApiUrl(`providers/${provider.spec.type}/${provider.metadata.uid}`),
+          );
+        }),
+    )
```
```suggestion
    const readyProviders = providers?.filter((provider) => provider.status.phase === 'Ready');
    const getInventoryURL(provider) => getInventoryApiUrl(`providers/${provider.spec.type}/${provider.metadata.uid}`);
        
    const allPromises = Promise.all(
      readyProviders.map(async (provider) => await consoleFetchJSON(getInventoryURL(provider)))
    )
```

yaacov commented on packages/forklift-console-plugin/src/modules/Providers/hooks/useProvidersInventoryList.ts:
```diff
@@ -67,7 +102,66 @@ export const useProvidersInventoryList = ({
+   * @param {V1beta1Provider[]} k8Providers providers to fetch inventory data for.
+   * @returns {void}
+   */
+  const getAndUpdateInventoryProviders = async (
+    k8Providers: V1beta1Provider[],
```
```suggestion
  const getInventoryByProviders = async (
    providers: V1beta1Provider[],
```

yaacov commented on packages/forklift-console-plugin/src/modules/Providers/hooks/useProvidersInventoryList.ts:
```diff
@@ -32,29 +39,42 @@ interface UseInventoryResult {
   const oldErrorRef = useRef(null);
+  const canList: boolean = useFlag('CAN_LIST_NS');
+
+  // Sanity check: if user does not have permissions to list all namespaces, exit now
+  if (!canList && namespace == null) {
+    const error = new Error("user doesn't have permissions to list all namespaces");
+    handleError(error);
+
+    return { inventory, loading: false, error };
+  }
```
we need to remove this test, because we do it before a hook ( useEffect )

sgratch replied:
I tried to only remove the `return` statement in line 69 but leave the error handling as is, since it won't change the main flow but will just set another more clear error message. But it was also rejected by the Code Analysis....

yaacov commented:
cypress test fail addressed by: #1317
